src/lib/storage/buffer/frame.cpp

```diff
@@ -0,0 +1,169 @@
+                                                      update_state_with_same_version(old_state_and_version, 1));
+  }
+  return false;
+}
+
+bool Frame::try_lock_exclusive(Frame::StateVersionType old_state_and_version) {
+  DebugAssert(state(old_state_and_version) == UNLOCKED || state(old_state_and_version) == MARKED ||
+                  state(old_state_and_version) == EVICTED,
+              "Frame must be unlocked, marked or evicted to lock exclusive, instead: " +
+                  std::to_string(state(old_state_and_version)));
+  return _state_and_version.compare_exchange_strong(old_state_and_version,
+                                                    update_state_with_same_version(old_state_and_version, LOCKED));
+}
+
+bool Frame::unlock_shared() {
+  while (true) {
+    auto old_state_and_version = _state_and_version.load();
+    auto old_state = state(old_state_and_version);
+    Assert(old_state > 0 && old_state <= LOCKED_SHARED, "Frame must be locked shared to unlock shared.");
+    auto new_state = old_state - 1;
+    if (_state_and_version.compare_exchange_strong(old_state_and_version,
+                                                   update_state_with_same_version(old_state_and_version, new_state))) {
+      return new_state == Frame::UNLOCKED;  // return true if last shared lock was released
+    }
+  }
+}
+
+void Frame::unlock_exclusive() {
+  Assert(state(_state_and_version.load()) == LOCKED,
+         "Frame must be locked to unlock exclusive. " + std::to_string(state(_state_and_version.load())));
+  _state_and_version.store(update_state_with_increment_version(_state_and_version.load(), UNLOCKED),
+                           std::memory_order_release);
+}
+
+Frame::StateVersionType Frame::update_state_with_same_version(Frame::StateVersionType old_version_and_state,
+                                                              Frame::StateVersionType new_state) {
+  constexpr auto SHIFT = NUM_BITS - STATE_SHIFT;
+  static_assert(SHIFT == 16, "Shift must be 8.");
+  return ((old_version_and_state << SHIFT) >> SHIFT) | (new_state << STATE_SHIFT);
+}
+
+Frame::StateVersionType Frame::update_state_with_increment_version(Frame::StateVersionType old_version_and_state,
+                                                                   Frame::StateVersionType new_state) {
+  constexpr auto SHIFT = NUM_BITS - STATE_SHIFT;
+  static_assert(SHIFT == 16, "Shift must be 8.");
+  return (((old_version_and_state << SHIFT) >> SHIFT) + 1) | (new_state << STATE_SHIFT);
+}
+
+bool Frame::is_unlocked() const {
+  return state(_state_and_version.load()) == UNLOCKED;
+}
+
+std::ostream& operator<<(std::ostream& os, const Frame& frame) {
+  auto state = Frame::state(frame.state_and_version());
+  auto version = Frame::version(frame.state_and_version());
+  auto node_id = frame.node_id();
+  auto dirty = frame.is_dirty();
+
+  std::stringstream ss;
+  ss << "Frame { state = ";
+
+  switch (state) {
```

Comment:
Seeing this code here: does it make sense to have the states in an enum with assigned values and cast or use the enum values in the code?

Reply:
I was doing this initially, too. However, you cannot represent values before `LOCKED_SHARED` properly.

Reply:
I think sth like
```c++
enum class FrameState : uint64_t { Unlocked = UNLOCKED, Locked = LOCKED, Marked = MARKED, Evicted = EVICTED };
```

and then sth like:
```c++
const auto state = magic_enum::enum_cast<FrameState>(...);
if (!state) {
  stream << "LockedShared";
} else {
   stream << magic_enum::enum_name(*state);
}
```
would work here. For other occasions, casting to the underlying type if necessary and comparing just like you did should also work. However, not a blocker.

Reply:
Yeah, but what’s the point of an enum if it doesn’t capture all values? Also, what happens if the state is larger than EVICTED?